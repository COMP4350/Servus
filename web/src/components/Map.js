import React, { useState, useEffect, useCallback } from 'react';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import { Button, TextField } from '@material-ui/core';
import { GoogleMap, useJsApiLoader } from '@react-google-maps/api';

const useStyles = makeStyles(theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-evenly',
    alignItems: 'center',
    margin: '10%'
  },
  textField : {
    width: '50%',
    marginBottom: '3%'
  },
  button : {
    marginBottom: '3%'
  }
}));

const mapStyle = {
  width: '400px',
  height: '400px'
};

const winnipeg = {
  latitude: 49.8951,
  longitude: 97.1384
};

const Location = () => {
  console.log(process.env.GOOGLE_MAPS_KEY);
  const classes = useStyles();
  const [center, setCenter] = useState();
  const { isLoaded } = useJsApiLoader({
    id: 'google-map-script',
    googleMapsApiKey: process.env.GOOGLE_MAPS_KEY
  });

  const [map, setMap] = useState(null);

  const onLoad = useCallback(map => {
    const bounds = new window.google.maps.LatLngBounds();
    map.fitBounds(bounds);
    setMap(map);
  }, []);

  useEffect(() => {
    if ("geolocation" in navigator) {
      setCenter(navigator.geolocation.getCurrentPosition(position => center = position));
    } else {
      center = winnipeg;
    }
  }, []);
  return (
    <div className={classes.root}>
      {isLoaded && (
        <GoogleMap
          mapContainerStyle={mapStyle}
          center = {center}
          zoom = {10}
          onLoad = {onLoad}/>
      )}
      <TextField className={classes.textField}
        id="outlined-basic"
        label="Location"
      />
    </div>
      
  );
};

export default Location;