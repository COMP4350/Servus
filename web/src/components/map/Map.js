import axios from 'axios';
import React, { useState, useCallback, useRef, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { TextField, IconButton } from '@material-ui/core';
import MyLocationIcon from '@material-ui/icons/MyLocation';
import {
    Autocomplete,
    GoogleMap,
    useJsApiLoader,
} from '@react-google-maps/api';
import mapStyle from './mapStyle.json';

const useStyles = makeStyles(() => ({
    root: {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        height: '100%',
        width: '100%',
    },
    textField: {
        backgroundColor: 'white',
    },
    addressContainer: {
        display: 'flex',
        justifyContent: 'center',
        marginTop: '2%',
        width: '100%',
    },
    mapContainer: {
        width: '100%',
        height: '90%',
        margin: 10,
    },
    button: {
        backgroundColor: 'white',
        position: 'absolute',
        bottom: '110px',
        right: '10px',
        borderRadius: '2px',
        padding: '8px',
        '&:hover': { backgroundColor: 'white' },
    },
    icon: {
        '&:hover': { color: 'black' },
    },
}));

const winnipeg = {
    lat: 49.8951,
    lng: -97.1384,
};

const mapLibraries = ['places'];

const defaultBounds = {
    north: winnipeg.lat + 1.0,
    south: winnipeg.lat - 1.0,
    east: winnipeg.lng + 1.0,
    west: winnipeg.lng - 1.0,
};

const autocompleteOptions = {
    bounds: defaultBounds,
    componentRestrictions: { country: 'ca' },
    fields: ['address_components', 'geometry', 'icon', 'name'],
    origin: winnipeg,
    strictBounds: true,
};

const mapOptions = {
    mapTypeControl: false,
    disableDefaultUI: true,
    zoomControl: true,
    styles: mapStyle,
};

const Map = () => {
    const classes = useStyles();
    const [center, setCenter] = useState(winnipeg);

    const { isLoaded } = useJsApiLoader({
        id: 'google-map-script',
        libraries: mapLibraries,
        googleMapsApiKey: process.env.REACT_APP_GOOGLE_MAPS_KEY,
    });

    let marker = useRef();
    let autocomplete = useRef();

    const onMapLoad = useCallback(map => {
        getUserLocation();
        marker.current = new window.google.maps.Marker({
            map,
            anchorPoint: new window.google.maps.Point(0, -29),
        });
        axios
            .get(`${process.env.REACT_APP_API_HOST}/services/`)
            .then(response => {
                let services = response.data.result;
                services &&
                    services.map(service => {
                        new window.google.maps.Marker({
                            map,
                            position: {
                                lat: service.location.lat,
                                lng: service.location.lng,
                            },
                            visible: true,
                        });
                    });
            });
    }, []);

    const onAutoCompleteLoad = useCallback(autocompleteLoaded => {
        autocomplete.current = autocompleteLoaded;
    }, []);

    const getUserLocation = () => {
        navigator.geolocation?.getCurrentPosition(position => {
            setCenter({
                lat: position.coords.latitude,
                lng: position.coords.longitude,
            });
        });
    };

    const onSearchAddressChanged = () => {
        setCenter(autocomplete.current?.getPlace().geometry?.location);
    };

    useEffect(() => {
        marker.current?.setPosition(center);
        marker.current?.setVisible(true);
    }, [center]);
    return (
        <div className={classes.root}>
            {isLoaded && (
                <GoogleMap
                    mapContainerClassName={classes.mapContainer}
                    center={center}
                    zoom={12}
                    mapTypeControl={false}
                    onLoad={onMapLoad}
                    options={mapOptions}>
                    <div className={classes.addressContainer}>
                        <Autocomplete
                            options={{ ...autocompleteOptions, origin: center }}
                            onPlaceChanged={onSearchAddressChanged}
                            onLoad={onAutoCompleteLoad}>
                            <TextField
                                className={classes.textField}
                                id="search-address"
                            />
                        </Autocomplete>
                    </div>
                    <IconButton
                        className={classes.button}
                        onClick={getUserLocation}>
                        <MyLocationIcon className={classes.icon} />
                    </IconButton>
                </GoogleMap>
            )}
        </div>
    );
};

export default Map;
