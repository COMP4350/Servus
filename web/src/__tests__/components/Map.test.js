import React from 'react';
import { act, render } from '@testing-library/react';
import Map from '../../components/map/Map';

const blankComponent = () => {
    return <div></div>;
};

jest.mock('@react-google-maps/api', () => {
    return {
        __esModule: true,
        useJsApiLoader: () => {
            return { isLoaded: true };
        },
        GoogleMap: () => {
            return blankComponent();
        },
        Autocomplete: () => {
            return blankComponent();
        },
    };
});
describe('Map', () => {
    it('renders correctly', async () => {
        const promise = Promise.resolve();

        const { asFragment } = render(<Map />);

        expect(asFragment()).toMatchSnapshot();

        await act(() => promise);
    });
});
