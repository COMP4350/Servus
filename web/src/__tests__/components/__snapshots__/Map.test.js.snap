// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1"
  >
    <div />
  </div>
</DocumentFragment>
`;
